test/stockpile_test.clj

```diff
@@ -0,0 +1,306 @@
+
+(deftest meta-encoding-round-trip
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [qdir (.toFile (.resolve tmpdir "queue"))
+           q (stock/create qdir)
+           batch-size 100]
+       (dotimes [i batch-size]
+         ;; We need to use a very short length here to avoid falling
+         ;; afoul of path length limits since 8 random unicode
+         ;; chars could expand to say 36 encoded bytes.
+         (let [meta (RandomStringUtils/random (rand-int 8))
+               entry (store-str q meta meta)]
+           (is (.startsWith meta (stock/entry-meta entry)))))))))
+
+(deftest existing-tmp-removal
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [qdir (.toFile (.resolve tmpdir "queue"))
+           garbage (File. qdir "q/tmp-garbage")]
+       (stock/create qdir)
+       (io/copy "foo" (File. qdir "q/tmp-garbage"))
+       (let [[q entries] (stock/open qdir conj ())]
+         (is (= [] entries))
+         (is (not (.exists garbage))))))))
+
+(deftest metadata-adjustment-failure
+  ;; Assumes that if "foo" fails, then "foo?" will fail too.
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [qdir (.toFile (.resolve tmpdir "queue"))
+           q (stock/create qdir)
+           retries (AtomicLong. 0)]
+       (try
+         (loop [i 100]
+           (store-str q
+                      "foo"
+                      (RandomStringUtils/randomAscii i)
+                      (fn [retry prev]
+                        (when (< retry 6)
+                          (.incrementAndGet retries)
+                          (str prev retry))))
+           (recur (* 2 i)))
+         (catch clojure.lang.ExceptionInfo ex
+           (let [data (ex-data ex)]
+             (is (= data {:kind :stockpile/commit-failed})))))
+       (is (= 5 (.get retries)))))))
+
+(deftest metadata-adjustment-success
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [qdir (.toFile (.resolve tmpdir "queue"))
+           q (stock/create qdir)
+           retries (AtomicLong. 0)
+           last-meta (atom nil)]
+       (loop [i 100]
+         (let [entry (store-str q
+                                "foo"
+                                (RandomStringUtils/randomAlphabetic i)
+                                (fn [retry prev]
+                                  (assert (< retries 10))
+                                  (.incrementAndGet retries)
+                                  (let [m (subs prev 0 (int (/ (count prev)
+                                                               2)))]
+                                    (reset! last-meta m)
+                                    m)))]
+           (if (zero? (.get retries))
+             (recur (* 2 i))
+             (do
+               (is (= "foo" (slurp-entry q entry)))
+               (is (= @last-meta (stock/entry-meta entry)))))))))))
+
+(def billion 1000000000)
+
+(defn enqueue-all [q items]
+  (doall (for [item items]
+           (apply store-str item))))
+
+(deftest uncontended-performance
```

Comment:
Somewhat, though note that these two tests (particularly the race) do demonstrate something if they succeed at all.  i.e. nothing caused a deadlock or threw an exception.   But of course the "rate" business is just informative right now.